sunpy/net/attr.py

```diff
@@ -23,16 +16,76 @@
 
-# XXX: Maybe allow other normal forms.
+_ATTR_TUPLE = namedtuple("attr", "name name_long desc")
+
+__all__ = ['Attr', 'DummyAttr', 'SimpleAttr', 'AttrAnd',
+           'AttrOr', 'ValueAttr', 'and_', 'or_']
+
+
+def make_tuple():
+    return _ATTR_TUPLE([], [], [])
+
 
+class AttrMeta(type):
+    """
+    We want to enable automatic discovery via tab completion of subclasses of Attrs.
+    To do this we have to create a metaclass that redefines the methods that Python uses to normally do this.
+    This would allow for example that `attrs.Instrument` to be able to tab complete to `attrs.Instrument.aia`.
+    However, it must be done by a downloader client which know what Attrs they support.
+    """
 
-class Attr(object):
+    # The aim is to register Attrs as a namedtuple of lists
+    # So we define the namedtuple outside of AttrMeta, see above.
+    _attr_registry = defaultdict(make_tuple)
+
+    def __getattr__(self, item):
+        """
+        Our method for Attrs is to register using `type(Attr)` as keys into a dictionary.
+        _attr_registry is a dictionary with a key being `type(Attr)` and the value being the namedtuple of lists.
+        As a result we index `_attr_registry` with `[self]` which will be the `type(Attr)` to access the dictionary.
+        This will return the namedtuple that has three attributes: `name`, `name_long` and `desc`. Each of which are a list.
+        `name` will be the attribute name, `name_long` is the original name passed in and `desc` the description of the object.
+        """
+        registry = self._attr_registry[self]  # Get the revelant entries.
+        names = registry.name  # All the attribute names under that type(Attr)
+        if item in names:
+            return self(registry.name_long[names.index(item)])  # We return Attr(name_long) to create the Attr requested.
+        else:
+            raise AttributeError('This attribute is not defined, please register it.')
+
+    def __dir__(self):
+        """
+        To tab complete in Python we need to add to the `__dir__()` return.
+        So we add the attribute name of any Attrs which have been added to the _attr_registry here.
+        """
+        return super().__dir__() + self._attr_registry[self].name
+
+    def __str__(self):
+        """
+        This enables the pretty printing of Attrs.
+        """
+        name = ['Attribute Name'] + self._attr_registry[self].name
+        name_long = ['Full Name'] + self._attr_registry[self].name_long
+        desc = ['Description'] + self._attr_registry[self].desc
+        pad_name = max(len(elm) for elm in name)
+        pad_long = max(len(elm) for elm in name_long)
+        pad_desc = max(len(elm) for elm in desc)
+        fmt = "%-{}s | %-{}s | %-{}s".format(pad_name, pad_long, pad_desc)
+        lines = [fmt % elm for elm in zip(name, name_long, desc)]
+        lines.insert(1, '-'*(pad_name+1)+'+'+'-'*(pad_long+2)+'+'+'-'*(pad_desc+1))
+        return("\n".join(lines))
```

Comment:
This entire block of code is ugly and I am sure could be cut down into something neat.

Reply:
Actually I think this is ok. You could probably shorten it but at the expense of readability.